compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt

```diff
@@ -0,0 +1,165 @@
+package com.squareup.anvil.compiler.dagger
+
+import com.google.common.truth.Truth.assertThat
+import com.squareup.anvil.compiler.USE_IR
+import com.squareup.anvil.compiler.WARNINGS_AS_ERRORS
+import com.squareup.anvil.compiler.internal.testing.compileAnvil
+import com.squareup.anvil.compiler.isFullTestRun
+import com.tschuchort.compiletesting.KotlinCompilation.ExitCode.COMPILATION_ERROR
+import com.tschuchort.compiletesting.KotlinCompilation.Result
+import org.intellij.lang.annotations.Language
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.junit.runners.Parameterized.Parameters
+
+@Suppress("UNCHECKED_CAST")
+@RunWith(Parameterized::class)
+class BindsMethodValidatorTest(
+  private val useDagger: Boolean
+) {
+
+  companion object {
+    @Parameters(name = "Use Dagger: {0}")
+    @JvmStatic fun useDagger(): Collection<Any> {
+      return listOf(isFullTestRun(), false).distinct()
+    }
+  }
+
+  @Test
+  fun `a binding with an incompatible parameter type fails to compile`() {
+    compile(
+      """
+      package com.squareup.test
+ 
+      import dagger.Binds
+      import dagger.Module
+      import javax.inject.Inject
+
+      class Foo @Inject constructor()
+      interface Bar
+
+      @Module
+      abstract class BarModule {
+        @Binds
+        abstract fun bindsBar(impl: Foo): Bar
+      }
+      """
+    ) {
+      assertThat(exitCode).isEqualTo(COMPILATION_ERROR)
```

Comment:
There's a shortcut: `assertThat(exitCode).isError()`. In all these instances you should also verify the error message to ensure that the correct error is printed. This is quite important.

Reply:
Good catch, updated now including error messages updated to exactly match Dagger's error messages.